src/compiler/checker.ts

```diff
@@ -27294,20 +27297,29 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
     /** remove undefined from the annotated type of a parameter when there is an initializer (that doesn't include undefined) */
     function removeOptionalityFromDeclaredType(declaredType: Type, declaration: VariableLikeDeclaration): Type {
-        if (pushTypeResolution(declaration.symbol, TypeSystemPropertyName.DeclaredType)) {
+        const links = getNodeLinks(declaration);
+
+        if (links.isParameterWithUndefinedInAnnotation === undefined) {
+            if (!pushTypeResolution(declaration, TypeSystemPropertyName.IsParameterWithUndefinedInAnnotation)) {
+                reportCircularityError(declaration.symbol);
+                return declaredType;
+            }
+
             const annotationIncludesUndefined = strictNullChecks &&
                 declaration.kind === SyntaxKind.Parameter &&
                 declaration.initializer &&
                 getTypeFacts(declaredType) & TypeFacts.IsUndefined &&
-                !(getTypeFacts(checkExpression(declaration.initializer)) & TypeFacts.IsUndefined);
-            popTypeResolution();
+                !(getTypeFacts(checkDeclarationInitializer(declaration, CheckMode.Normal)) & TypeFacts.IsUndefined);

```

Comment:
Alternatively, I could avoid setting this on most nodes (only applies to parameters) and skip this circularity, but that felt dangerous.

Reply:
Nah, this function simply won't be called on anything but params already, so id optimized enough .